File: SystemDesign/ObserverDesignPattern/odp.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>

using namespace std;

// Observer Interface
class ISubscriber {
public:
    virtual void update(const string& videoTitle,
                        const string& channelName) = 0;
    virtual ~ISubscriber() {}
};

// Subject Interface
class IChannel {
public:
    virtual void subscribe(ISubscriber* subscriber) = 0;
    virtual void unsubscribe(ISubscriber* subscriber) = 0;
    virtual void notifySubscribers() = 0;
    virtual ~IChannel() {}
};

// Concrete Subject
class Channel : public IChannel {
private:
    vector<ISubscriber*> subscribers;
    string name;
    string latestVideo;

public:
    Channel(const string& name) : name(name) {}

    void subscribe(ISubscriber* subscriber) override {
        if (find(subscribers.begin(), subscribers.end(), subscriber) == subscribers.end()) {
            subscribers.push_back(subscriber);
        }
    }

    void unsubscribe(ISubscriber* subscriber) override {
        auto it = find(subscribers.begin(), subscribers.end(), subscriber);
        if (it != subscribers.end()) {
            subscribers.erase(it);
        }
    }

    void notifySubscribers() override {
        for (auto subscriber : subscribers) {
            subscriber->update(latestVideo, name);
        }
    }

    void uploadVideo(const string& title) {
        latestVideo = title;

        cout << "\n====================================\n";
        cout << name << " uploaded: " << latestVideo << endl;
        cout << "====================================\n";

        notifySubscribers();
    }
};

// Concrete Observer
class Subscriber : public ISubscriber {
private:
    string name;

public:
    Subscriber(const string& name) : name(name) {}

    void update(const string& videoTitle,
                const string& channelName) override {
        cout << "Hey " << name
             << ", " << channelName
             << " uploaded a new video: "
             << videoTitle << endl;
    }
};

int main() {

    Channel coderArmy("CoderArmy");

    // No relationship with any channel yet
    Subscriber varun("Varun");
    Subscriber tarun("Tarun");

    // Relationship starts here
    coderArmy.subscribe(&varun);
    coderArmy.subscribe(&tarun);

    coderArmy.uploadVideo("Observer Pattern Tutorial");

    coderArmy.unsubscribe(&varun);

    coderArmy.uploadVideo("Decorator Pattern Tutorial");

    return 0;
}

File: SystemDesign/ObserverDesignPattern/ODP.py
from abc import ABC,abstractmethod

class Ichannel:
    @abstractmethod
    def subscribe(self):
        pass

    @abstractmethod
    def unsubscribe(self):
        pass

    @abstractmethod
    def notify(self):
        pass

class Isubscriber:
    @abstractmethod
    def update(self):
        pass

class Channel(Ichannel):
    def __init__ (self,name):
        self.name = name
        self.latest_video = None
        self.subscribers = {}
    def subscribe(self,user_name):
        if user_name not in self.subscribers:
            self.subscribers[user_name] = True

    def unsubscribe(self, user_name):
        if user_name in self.subscribers:
            del self.subscribers[user_name]

    def notify(self):
        for key,val in self.subscribers.items():
            key.update(self.latest_video,self.name)

    def upload_video(self,video_title):
        self.latest_video = video_title
        self.notify()


class Subscriber(Isubscriber):
    def __init__ (self,name):
        self.name = name


    def update(self,video_name,channel_name):
        print("Hi",self.name)
        print("From :",channel_name)
        print("New video Uploaded :", video_name)
        print("="*40)
        

channel = Channel("Marvel")

user1 = Subscriber("Sameer")
user2 = Subscriber("Rishu")

channel.subscribe(user1)
channel.subscribe(user2)

channel.upload_video("Spider Man : Brand New Day")

channel.unsubscribe(user2)
channel.upload_video("Avengers : DoomsDay")